### DISCUSSION TITLE
SegmentedControl

### DISCUSSION THREAD

**Discussion by kazo0** [unoplatform/uno#6067] [Ideas]
![image](https://user-images.githubusercontent.com/4793020/119495335-5067ec80-bd30-11eb-9992-0c81565deded.png)

There is a need to create a control similar to what is being discussed here: https://github.com/microsoft/microsoft-ui-xaml/issues/2310. 

A very basic implementation may be just a re-styling of [RadioButtons](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.radiobuttons?view=winui-3.0) but that may not be enough since we would want to support things like multiple selection and gestures such as swiping with a sliding selection indicator. 

> Could this be a restyle of RadioButtons? Likely not if we want to support multiple selection, swipe, carouseling, or other more complicated functionality

For a "sliding" segmented control as mentioned above, we would want to achieve something similar to Flutter's [CupertinoSlidingSegmentedControl](https://api.flutter.dev/flutter/cupertino/CupertinoSlidingSegmentedControl-class.html) as seen in the [Flutter Gallery](https://gallery.flutter.dev/#/demo/cupertino-segmented-control):
![sliding-segmented](https://user-images.githubusercontent.com/4793020/119497813-fae10f00-bd32-11eb-896e-0510d4f7a2e1.gif)

# Implementation
Rather than just a re-style of the RadioButtons control, I think we should be creating a new control using an [ItemsRepeater](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/items-repeater) as mentioned in https://github.com/microsoft/microsoft-ui-xaml/issues/2310#issuecomment-800495012:

> I don't think this should actually be implemented with ToggleButtons or RadioButtons.
This whole control could be internally implemented using ItemsRepeater, SelectionModel and the attached layout system. UniformGridLayout and StackLayout can already handle all the layout cases I described above. We would essentially just be creating nice wrapping over this architecture to make it more developer friendly and calling it a ButtonGroup. The RadioButtons control did something similar I believe.

We would then implement the same model of `SelectedItems` that [ListView](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.listview?view=winui-3.0) has, with a `SelectionMode`. 

The split here is how this will be implemented for single-selection and multi-selection. If we are in `SelectionMode.Single`, then we could technically just be holding an `ItemsRepeater` of styled `RadioButton`s. If we are in `SelectionMode.Multiple` then we could be holding an `ItemsRepeater` of `ToggleButton`s.

**Comment by kazo0** [comment]
Another option that we can go for is just to create a custom style for the `TabBar`. `TabBar` already has the ability to have a `TabBarSelectionIndicatorPresenter` added to its Control Template.

**Reply by robloo** [reply]
`TabBar` implies single selection and navigation usage. This control has optional multi-selection and settings usages as well. I would definitely recommend a new control for this.

**Comment by robloo** [comment]
I had another thought about this: its probably better to implement this in the Windows Community Toolkit first. It will get a bit more engagement from the full community and can quickly be ported to other frameworks as needed (license is MIT).

**Reply by francoistanguay** [reply]
Agreed. That's why for now, the only thing we'll do is style our upcoming TabBar control with a style for https://gallery.flutter.dev/#/demo/cupertino-segmented-control. Our TabBar is also being given a Material style as in https://material.io/components/tabs#fixed-tabs
